### PR TITLE
Add image tag functionality

### LINKED ISSUE #7: Image Tags
#### As a user I want to add tags to images.

Acceptance criteria:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image on the index page.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

Reference material:
- https://github.com/mbleigh/acts-as-taggable-on

Dependencies:
- Image Index

### LINKED ISSUE #9: Images by Tag
#### As a user I want to view all images associated with a tag.

Acceptance criteria:
- [x] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [x] The tag name (not its ID) is used to perform the filtering.

Discussion topics:
- [x] Where should we implement this functionality?
- [x] What happens if a user tries to manually filter by a nonexistent tag?

Dependencies:
- Image Tags

### PR DESCRIPTION
**Pull request by wpliao1989**
Closes #7

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  # rubocop:disable Metrics/MethodLength
+  def self.up
+    create_table :tags do |t|
+      t.string :name
+    end
+
+    create_table :taggings do |t|
+      t.references :tag
```
Comment (shfung) [inline db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Maybe we can add some not-null constraints to some of the columns?

Comment (wpliao1989) [inline db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: I don't know how the gem manages these tables but I'm all for it.

File: app/views/images/_image_tags.erb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<% image.tags.any? do %>
```
Comment (shfung) [inline app/views/images/_image_tags.erb]: Can you add the `html` extension before the `.erb`?

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -17,13 +20,16 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     assert_select 'form'
   end
 
-  test 'should create with valid url' do
-    url = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
+  test 'should create' do
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Can you add a test for the `create` action with empty tag?

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -17,13 +20,16 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+      post images_url image: { url: url, tag_list: tags }
     end
     assert_redirected_to image_url(Image.last)
-    assert_equal 'New image created!', flash[:notice]
+    assert_equal 'New image created!', flash[:success]
+    image = Image.last
+    assert_equal %w(cats dogs), image.tag_list
   end
 
   test 'should not create with invalid url' do
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Can you also add the key for tag_list in this test?

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -37,10 +43,13 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
 
   test 'should get show' do
     url   = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
-    image = Image.create(url: url)
+    tags  = 'cats, dogs'
+    image = Image.create(url: url, tag_list: tags)
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Maybe use `create!` instead of `create` so that this line will break if the `Image` cannot be created?

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -2,13 +2,16 @@
 
 class ImagesControllerTest < ActionDispatch::IntegrationTest
   test 'should get index' do
-    url = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
-    Image.create(url: url)
+    url  = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
+    tags = 'cats, dogs'
     Image.create(url: url)
+    Image.create(url: url, tag_list: tags)
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Same comment as above: Maybe use `create!` instead of `create` so that this line will break if the `Image` cannot be created?

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -17,13 +20,16 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+    url  = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
+    tags = 'cats, dogs'
     assert_difference('Image.count') do
-      post images_url image: { url: url }
+      post images_url image: { url: url, tag_list: tags }
     end
     assert_redirected_to image_url(Image.last)
-    assert_equal 'New image created!', flash[:notice]
+    assert_equal 'New image created!', flash[:success]
+    image = Image.last
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Maybe move this line before the `assert_redirected_to` and use `image` there so that you don't issue two queries to the db?

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -37,10 +43,13 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
 
   test 'should get show' do
     url   = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
-    image = Image.create(url: url)
+    tags  = 'cats, dogs'
+    image = Image.create(url: url, tag_list: tags)
     get image_url id: image.id
     assert_response :success
-    assert_select 'p'
+    assert_select 'img'
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: The assertions in this test is the same as the assertions in the `should get index` test?

**Comment by shfung** [comment]
You should also add unit tests for the model and flow tests.

**Comment by wpliao1989** [comment]
Got it, I'll work on those, thanks!

**Comment by wpliao1989** [comment]
Updated

File: test/controllers/images_controller_test.rb (hunk 7 above)
Comment (shfung) [inline test/controllers/images_controller_test.rb]: This comment is not addressed yet.

File: test/flows/image_flows_test.rb [hunk 9]
```diff
@@ -1,19 +1,35 @@
...
 
+  # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
+  def create_image_test_helper(url, tags)
+    visit new_image_path
     fill_in('Url', with: 'test')
+    fill_in('Tag list', with: 'tags')
     click_button('Create Image')
-    assert_selector('.error', text: 'is invalid')
+    assert_selector('.error', count: 1, text: 'is invalid')
     assert page.has_current_path?(images_path)
```
Comment (shfung) [inline test/flows/image_flows_test.rb]: I think the above error test case belongs to a separate test rather than doing it twice in both cases.

File: test/controllers/images_controller_test.rb (hunk 7 above)
Comment (wpliao1989) [inline test/controllers/images_controller_test.rb]: I did one but not the other, thanks for catching this!

File: test/flows/image_flows_test.rb (hunk 9 above)
Comment (wpliao1989) [inline test/flows/image_flows_test.rb]: You're right, splitting them makes it more clear.

**Comment by wpliao1989** [comment]
Updated

File: app/views/images/_image_tags.html.erb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<% image.tags.any? do %>
```
Comment (shfung) [inline app/views/images/_image_tags.html.erb]: While this current implementation works, I think it'll be easier to understand if you do something like:
```
<% if image.tags.present? %>
```

Comment (wpliao1989) [inline app/views/images/_image_tags.html.erb]: Hmm that's really weird because it works. I'll look into it.

File: test/flows/image_flows_test.rb [hunk 11]
```diff
@@ -1,19 +1,39 @@
...
+  test 'create a new image with non empty tags' do
+    url = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
+    tags = 'cats, dogs'
     visit new_image_path
+    test_valid_form url, tags
+  end
+
+  private
 
+  def test_invalid_form
```
Comment (shfung) [inline test/flows/image_flows_test.rb]: Minor: while private methods are not considered as tests, maybe we should avoid naming them like `test_<blah>` to avoid confusions?

File: test/flows/image_flows_test.rb [hunk 12]
```diff
@@ -1,19 +1,39 @@
...
     fill_in('Url', with: 'test')
+    fill_in('Tag list', with: 'tags')
     click_button('Create Image')
-    assert_selector('.error', text: 'is invalid')
+    assert_selector('.error', count: 1, text: 'is invalid')
     assert page.has_current_path?(images_path)
+  end
 
+  # rubocop:disable Metrics/AbcSize
+  def test_valid_form(url, tags)
```
Comment (shfung) [inline test/flows/image_flows_test.rb]: Minor: while private methods are not considered as tests, maybe we should avoid naming them like `test_<blah>` to avoid confusions?

File: test/models/image_test.rb [hunk 13]
```diff
@@ -20,6 +20,37 @@ def test_url_invalid
...
+    valid_url = valid_urls.first
+    tags = 'cats, , dogs, '
+    image = Image.new(url: valid_url, tag_list: tags)
+    assert_predicate image, :valid?
+    assert_equal %w(cats dogs), image.tag_list
+  end
+
   private
 
   def test_url_invalid_helper(image, message = 'is invalid')
```
Comment (shfung) [inline test/models/image_test.rb]: Minor: while private methods are not considered as tests, maybe we should avoid naming them like `test_<blah>` to avoid confusions?

Comment (wpliao1989) [inline test/models/image_test.rb]: Yes makes sense. Never thought of that.

**Comment by wpliao1989** [comment]
Updated

**Comment by shfung** [comment]
Looks good! Can you combine the `Add db constraint` commit with the `Add image tag functionality` commit so that a migration file is contained within a single commit and is not changed across commits?

**Comment by wpliao1989** [comment]
Cool! I'll squash them and force push.

**Comment by wpliao1989** [comment]
Done! Shall I go ahead and merge it?

**Comment by shfung** [comment]
👍 !

File: circle.yml [hunk 14]
```diff
@@ -1,3 +1,6 @@
 test:
   post:
     - bundle exec rubocop
+database:
+  override:
+    - bundle exec rails db:create db:migrate RAILS_ENV=test
```
Comment (bboe) [inline circle.yml]: It's frustrating that this is necessary.

File: db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+      t.string :name, null: false
+    end
+
+    create_table :taggings do |t|
+      t.references :tag, null: false
+
+      # You should make sure that the column created is
+      # long enough to store the required class names.
+      t.references :taggable, polymorphic: true, null: false
+      t.references :tagger, polymorphic: true
```
Comment (bboe) [inline db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: For future reference, even though `null: true` is the default, I would say it's a good idea to be explicit about fields that can be nullable. Otherwise, I as a code reviewer, might think you didn't consider it.

Comment (wpliao1989) [inline db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: I agree

File: circle.yml (hunk 14 above)
Comment (wpliao1989) [inline circle.yml]: Haven't looked deeply into it yet.. your fix somehow didn't work so I'm just sticking to what works. This looks hacky lol.

File: db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 15 above)
Comment (wpliao1989) [inline db/migrate/20160726175534_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Good advice!
